libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,2036 @@
+        in IotHttpClient_InitializeRequest(). */
+        if( strncmp(pName, HTTPS_HOST_HEADER, strlen(HTTPS_HOST_HEADER)) == 0)
+        {
+            IotLogError("Attempting to add auto-generated header %s. This is not allowed.", HTTPS_HOST_HEADER);
+            status = IOT_HTTPS_INVALID_PARAMETER;
+        }
+    }
+
+    if(status == IOT_HTTPS_OK)
+    {
+        /* Check for auto-generated header "User-Agent". This header is created and placed into the header buffer space
+        in IotHttpClient_InitializeRequest(). */
+        if(strncmp(pName, HTTPS_USER_AGENT_HEADER, strlen(HTTPS_USER_AGENT_HEADER)) == 0)
+        {
+            IotLogError("Attempting to add auto-generated header %s. This is not allowed.", HTTPS_USER_AGENT_HEADER);
+            status = IOT_HTTPS_INVALID_PARAMETER;
+        }
+    }
+
+    if(status == IOT_HTTPS_OK)
+    {
+        status = _addHeader(reqHandle, pName, pValue, len );
+        if(status != IOT_HTTPS_OK)
+        {
+            IotLogError("Error in IotHttpsClient_AddHeader(), error code %d.", status);
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _networkSend(_httpsConnection_t* _httpsConnection, uint8_t * pBuf, size_t len)
+{
+    size_t numBytesSent = 0;
+    size_t numBytesSentTotal = 0;
+    size_t sendLength = len;
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    /* Send the headers first because the HTTPS body is in a separate pointer these are not contiguous. */
+    while( numBytesSentTotal < sendLength )
+    {
+        numBytesSent = _httpsConnection->pNetworkInterface->send( _httpsConnection->pNetworkConnection, 
+            &( pBuf[numBytesSentTotal] ), 
+            sendLength - numBytesSentTotal );
+
+        if( numBytesSent <= 0 )
+        {
+            IotLogError("Error in sending the HTTPS headers. Error code: %d", numBytesSent);
+            break;
+        }
+
+        numBytesSentTotal += numBytesSent;
+    }
+
+    if( numBytesSentTotal != sendLength )
+    {
+        IotLogError("Error sending data on the network. We sent %d but there is %d left to send.", numBytesSentTotal, sendLength);
+        status = IOT_HTTPS_NETWORK_ERROR;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _networkRecv( _httpsConnection_t* _httpsConnection, uint8_t * pBuf, size_t bufLen)
+{
+    size_t numBytesRecv = 0;
+    size_t numBytesRecvTotal = 0;
+    size_t lengthToReceive = bufLen;
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    do {
+        numBytesRecv = _httpsConnection->pNetworkInterface->receive( _httpsConnection->pNetworkConnection,
+            &( pBuf[ numBytesRecvTotal ] ),
+            lengthToReceive - numBytesRecvTotal );
+
+        if( numBytesRecv > 0 )
+        {
+            numBytesRecvTotal += numBytesRecv;
+        }
+        if( numBytesRecv == 0)
```

Comment:
This could also mean that server has closed the connection.  The client should close the connection in this case.  Please make if calling function closes the the connection.

Reply:
If the server closed the connection, then a negative error code is returned. I have experimented with this case. I cannot close for every timeout because I am need to try and get all data from the socket without knowing how much there is.

Reply:
Returning zero does indicate a socket closed the connection: See last line here:
```
        /* Block and wait for 1 byte of data. This simulates the behavior of poll().
         * THIS IS A TEMPORARY WORKAROUND AND DOES NOT PROVIDE THREAD-SAFETY AGAINST
         * MULTIPLE CALLS OF RECEIVE. */
        do
        {
            socketStatus = SOCKETS_Recv( pNetworkConnection->socket,
                                         &( pNetworkConnection->bufferedByte ),
                                         1,
                                         0 );

            connectionFlags = xEventGroupGetBits( ( EventGroupHandle_t ) &( pNetworkConnection->connectionFlags ) );

            if( ( connectionFlags & _FLAG_SHUTDOWN ) == _FLAG_SHUTDOWN )
            {
                socketStatus = SOCKETS_ECLOSED;
            }

            /* Check for timeout. Some ports return 0, some return EWOULDBLOCK. */
        } while( ( socketStatus == 0 ) || ( socketStatus == SOCKETS_EWOULDBLOCK ) );
```